OptimizelySDK/Odp/OdpEventManager.cs

```diff
@@ -0,0 +1,418 @@
+        {
+            _logger.Log(LogLevel.DEBUG, "Stop requested.");
+
+            // process queue with flush
+            ProcessQueue(true);
+
+            State = ExecutionState.Stopped;
+            _logger.Log(LogLevel.DEBUG, $"Stopped. Queue Count: {_queue.Count}.");
+        }
+
+        /// <summary>
+        /// Register a new visitor user id (VUID) in ODP
+        /// </summary>
+        /// <param name="vuid">Visitor ID to register</param>
+        public void RegisterVuid(string vuid)
+        {
+            var identifiers = new Dictionary<string, string>
+            {
+                {
+                    OdpUserKeyType.VUID.ToString(), vuid
+                },
+            };
+
+            var odpEvent = new OdpEvent(TYPE, "client_initialized", identifiers);
+            SendEvent(odpEvent);
+        }
+
+        /// <summary>
+        /// Associate a full-stack userid with an established VUID
+        /// </summary>
+        /// <param name="userId">Full-stack User ID</param>
+        /// <param name="vuid">Visitor User ID</param>
+        public void IdentifyUser(string userId, string vuid)
+        {
+            var identifiers = new Dictionary<string, string>
+            {
+                {
+                    OdpUserKeyType.FS_USER_ID.ToString(), userId
+                },
+            };
+
+            if (!string.IsNullOrWhiteSpace(vuid))
+            {
+                identifiers.Add(OdpUserKeyType.VUID.ToString(), vuid);
+            }
+
+            var odpEvent = new OdpEvent(TYPE, "identified", identifiers);
+            SendEvent(odpEvent);
+        }
+
+        /// <summary>
+        /// Send an event to ODP via dispatch queue
+        /// </summary>
+        /// <param name="odpEvent">ODP Event to forward</param>
+        public void SendEvent(OdpEvent odpEvent)
+        {
+            if (State == ExecutionState.Stopped)
+            {
+                _logger.Log(LogLevel.WARN,
+                    "Failed to Process ODP Event. ODPEventManager is not running.");
+                return;
+            }
+
+            if (!IsOdpConfigurationReady())
+            {
+                return;
+            }
+
+            if (InvalidDataFound(odpEvent.Data))
+            {
+                _logger.Log(LogLevel.ERROR, "Event data found to be invalid.");
+            }
+            else
+            {
+                odpEvent.Data = AugmentCommonData(odpEvent.Data);
+                Enqueue(odpEvent);
+            }
+        }
+
+        /// <summary>
+        /// Add a new ODP event to the queue
+        /// </summary>
+        /// <param name="odpEvent"></param>
+        private void Enqueue(OdpEvent odpEvent)
+        {
+            if (_queue.Count >= _queueSize)
+            {
+                _logger.Log(LogLevel.WARN,
+                    $"Failed to Process ODP Event. Event Queue full. queueSize = {_queue.Count}.");
+                return;
+            }
+
+            _queue.Enqueue(odpEvent);
```

Comment:
Is this thread-safe?

Reply:
Yes.  The ConcurrentQueue should handle all CRUD operations in a thread safe manner.  We get to interact with it in the familiar collection-y way.